### PR TITLE
Simple Texting Webhook

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

Added webhook for Simple Texting. Supports delivery confirmation and processing incoming messages for sms pipeline, standard processing and sms conversations. Also supports media/mms attachments for v13+

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

- Added webhook for SimpleTexting. Supports delivery confirmation and processing incoming messages for sms pipeline, standard processing and sms conversations. Also supports media/mms attachments for v13+

---------

##### Who reported, requested, or paid for the change?

VWC

---------

##### Does this update or add options to the block UI?

Internal to Rock

---------

##### What files does it affect?

Webhooks/SimpleTexting.ashx

---------

##### Is it a breaking change for other versions/clients?

Requires the communication transport/component from https://github.com/KingdomFirst/RockAssemblies/pull/103

### CODE REVIEW

File: Webhooks/SimpleTexting.ashx [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+                // Simple Texting does not just respond to text messages based on the response in the webhook, so we need to standup the component, client and actually send a message with the methods we have.
+                var simpleTextingComponent = new rocks.kfs.SimpleTexting.Communications.Transport.SimpleTexting();
+                simpleTextingComponent.LoadAttributes();
+
+                if ( simpleTextingComponent.IsActive )
+                {
+                    simpleTextingClient = new SimpleTextingDotNet.v2.Client( simpleTextingComponent.GetAttributeValue( rocks.kfs.SimpleTexting.Communications.Transport.SimpleTexting.AttributeKey.ApiKey ) );
+                }
+
+                if ( messageReport != null && messageReport.Values != null && messageReport.Values.MediaItems != null && messageReport.Values.MediaItems.Any() )
```
Comment (nateh777) [inline Webhooks/SimpleTexting.ashx:388]:
```suggestion
                if ( messageReport != null && messageReport.Values != null && messageReport.Values.MediaItems != null && messageReport.Values.MediaItems.Any() && simpleTextingClient != null )
```

File: Webhooks/SimpleTexting.ashx [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+                        attachmentUrls.Add( binaryFile.Url );
+                    }
+                }
+
+                if ( !string.IsNullOrWhiteSpace( smsResponse.Message ) )
+                {
+                    if ( simpleTextingClient != null )
+                    {
+                        simpleTextingClient.SendMessage( message.FromNumber, smsResponse.Message, accountPhone: message.ToNumber, mediaItems: attachmentUrls );
+                        return true;
```
Comment (nateh777) [inline Webhooks/SimpleTexting.ashx:459]:
```suggestion
                        return true; // Message was sent back.
```

File: Webhooks/SimpleTexting.ashx [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+            var processedObj = ProcessJsonContent( request, response );
+            if ( processedObj != null )
+            {
+                switch ( processedObj.GetType().Name )
+                {
+                    case nameof( UnsubscribeReport ):
+                        // ignore Unsubscribe reports for now, perhaps if we add contact syncing or group sync with contact lists we will remove people here.
+                        break;
+                    case nameof( MessageReport ):
+                        var messageReport = ( MessageReport ) processedObj;
```
Comment (guymaness2) [inline Webhooks/SimpleTexting.ashx:152]: Should probably check that the processedObj values are valid as well.

File: Webhooks/SimpleTexting.ashx [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+                    ExceptionLogService.LogException( $"No recipient was found with the specified MessageId value! ({messageId})" );
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Mark message as delivered.
+    /// </summary>
+    private void MessageDelivered( DeliveryMessageReport deliveryReport )
```
Comment (guymaness2) [inline Webhooks/SimpleTexting.ashx:277]: Looks like a lot of shared code with the method above. Can they be simplified to a single method?

File: Webhooks/SimpleTexting.ashx (hunk 3 above)
Comment (nateh777) [reply Webhooks/SimpleTexting.ashx:152]: I believe the checks in each sub-method are sufficient. I did look into newtonsoft validation method, they can validate json to be serialized but not the other direction.

File: Webhooks/SimpleTexting.ashx [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+    /// <summary>
+    /// Gets the type of the file extension from content. If you need to go the other way use System.Web.MimeMapping.GetMimeMapping(String)
+    /// </summary>
+    /// <returns></returns>
+    public static string GetFileExtensionFromContentType( string contentType )
+    {
+        // Add to this list as needed.
+        var mapping = new Dictionary<string, string>( StringComparer.OrdinalIgnoreCase )
+            {
+                { "bmp", "image/bmp" },
```
Comment (guymaness2) [inline Webhooks/SimpleTexting.ashx:482]: Is this list of supported file types extensive enough?

**Pull request merged**